text/0000-try-trait.md

```diff
@@ -0,0 +1,553 @@
+    Ok(T),
+    Abrupt(E),
+}
+```
+
+Re-using `Result` was chosen because it is simpler (fewer things being
+added).  It also allows manual invocations of `Try` (should there by
+any, perhaps in macros) to re-use the rich set of methods available on
+`Result`. Finally, the `Result` type seems semantically
+appropriate. In general, `Result` is used to indicate whether normal
+execution can continue (`Ok`) or whether some form of recoverable
+error occurred (`Err`).  In this case, "normal execution" means the
+rest of the function, and the recoverable error means abrupt
+termination:
+
+- returning a `Ok(v)` value means that (a) execution
+  should continue normally and (b) the result `v` represents the value to
+  be propagated;
+- returning a `Err(e)` value means that (a) execution should return
+  abruptly and (b) the result `e` is the value to be propagated (note
+  that `e` may itself be a `Result`, if the function returns
+  `Result`).
+  
+### The original `Carrier` trait proposal
+
+The original `Carrier` as proposed in RFC 243 had a rather different
+design:
+
+```rust
+trait ResultCarrier {
+    type Normal;
+    type Exception;
+    fn embed_normal(from: Normal) -> Self;
+    fn embed_exception(from: Exception) -> Self;
+    fn translate<Other: ResultCarrier<Normal=Normal, Exception=Exception>>(from: Self) -> Other;
+}
+```
+
+Whereas this `Try` trait links the type of the value being matched
+(`Self`) with the type that the enclosing function returns (`E`), the
+`Carrier` was implemented separately for each type (e.g., `Result` and
+`Option`). It allowed any kind of carrier to be converted to any other
+kind of carrier, which fails to preserve the "semantic equivalent"
+property.  It would also interact poorly with type inference, as
+discussed in the "Detailed Design" section.
+
+### Traits implemented over higher-kinded types
+
+The "semantic equivalent" property might suggest that the `Carrier`
+trait ought to be defined over higher-kinded types (or generic
+associated types) in some form. The most obvious downside of such a
+design is that Rust does not offer higher-kinded types nor anything
+equivalent to them today, and hence we would have to block on that
+design effort. But it also turns out that HKT is
+[not a particularly good fit for the problem](https://github.com/rust-lang/rust/pull/35056#issuecomment-240129923). To
+start, consider what "kind" the `Self` parameter on the `Try` trait
+would have to have.  If we were to implement `Try` on `Option`, it
+would presumably then have kind `type -> type`, but we also wish to
+implement `Try` on `Result`, which has kind `type -> type ->
+type`. There has even been talk of implementing `Try` for simple types
+like `bool`, which simply have kind `type`. More generally, the
+problems encountered are quite similar to the problems that
+[Simon Peyton-Jones describes in attempting to model collections using HKT](https://github.com/rust-lang/rust/pull/35056#issuecomment-240129923):
```

Comment:
That link seems to go to the wrong place... at least, the referenced comment is neither written by Simon Peyton-Jones nor is it about collections. ;)